eth/filters/api.go

```diff
@@ -280,8 +321,178 @@ func (api *FilterAPI) Logs(ctx context.Context, crit FilterCriteria) (*rpc.Subsc
 			}
 		}
 	}()
+	return nil
+}
 
-	return rpcSub, nil
+// histLogs retrieves logs older than current header.
+func (api *FilterAPI) histLogs(notifier notifier, rpcSub *rpc.Subscription, from int64, crit FilterCriteria) error {
+	// Subscribe the Live logs
+	// if an ChainReorg occurred,
+	// we will first recv the old chain's deleted logs in descending order,
+	// and then the new chain's added logs in descending order
+	// see core/blockchain.go#reorg(oldHead *types.Header, newHead *types.Block) for more details
+	// if an reorg happened between `from` and `to`, we will need to think about some scenarios:
+	// 1. if a reorg occurs after the currently delivered block, then because this is happening in the future, has nothing to do with the current historical sync, we can just ignore it.
+	// 2. if a reorg occurs before the currently delivered block, then we need to stop the historical delivery, and send all replaced logs instead
+	var (
+		liveLogs = make(chan []*types.Log)
+		histLogs = make(chan []*types.Log)
+		histDone = make(chan error)
+	)
+	liveLogsSub, err := api.events.SubscribeLogs(ethereum.FilterQuery(crit), liveLogs)
+	if err != nil {
+		return err
+	}
+
+	// The original request ctx will be canceled as soon as the parent goroutine
+	// returns a subscription. Use a new context instead.
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		histDone <- api.doHistLogs(ctx, from, crit.Addresses, crit.Topics, histLogs)
+	}()
+
+	// Compose and notify the logs from liveLogs and histLogs
+	go func() {
+		defer func() {
+			liveLogsSub.Unsubscribe()
+			cancel()
+		}()
+		var (
+			delivered uint64
+			liveOnly  bool
+			reorged   bool
+			staleHash common.Hash
+		)
+		for {
+			select {
+			case err := <-histDone:
+				if err != nil {
+					logger.Warn("History logs delivery failed", "err", err)
+					return
+				}
+				// Else historical logs are all delivered, let's switch to live mode
+				logger.Info("History logs delivery finished, and now enter into live mode", "delivered", delivered)
+				liveOnly = true
+				histLogs = nil
+
+			case logs := <-liveLogs:
+				if len(logs) == 0 {
+					continue
+				}
+				if liveOnly {
+					notifyLogsIf(notifier, rpcSub.ID, logs, nil)
+					continue
+				}
+				reorgBlock := logs[0].BlockNumber
+				if !reorged && reorgBlock <= delivered {
+					logger.Info("Reorg detected", "reorgBlock", reorgBlock, "delivered", delivered)
+					reorged = true
```

Comment:
This seems a bit strange. Once `reorged` i sset, it never becomes unset again. It will discard all `histLogs` when it's in that mode. Seems to me that this big switch has thee different modes: 

- reorged mode, where it discards historic logs
- liveOnly mode
- "normal" non-reorged mode

It makes it a bit complicated to follow, IMO

Reply:
> Once reorged i sset, it never becomes unset again

Here we only need to handle the case when our historical delivery process is catching up with the live blocks.

> liveOnly mode

Yeah, We did have another switch `liveOnly`, which was set after all historical logs were delivered: https://github.com/ethereum/go-ethereum/blob/633c7ac67a6de35e000bdc72d51f1054f4e743b5/eth/filters/api.go#L373-L375

Reply:
> Seems to me that this big switch has thee different modes:

Actually I managed to bring it down to 2 modes (https://github.com/ethereum/go-ethereum/pull/27439/commits/674a888b060b6acabff0cdcd3ab12fc8fbecb5fd). Now only setting `liveOnly` mode. This will be set:

- either when a reorg happens to discard historical logs
- or when historical logs have finished

But my patch has a problem. In the "reorg" mode, we only send removed log notifications up to the `delivered` point. So if we finish historical processing (say head = 10 so delivered = 10). Then after emitting block 11, we get a reorg from block 8. We will only delivered the removed logs up to 10.

Reply:
Ok I pushed another commit which should fix the problem above and IMO the code is more readable now.

That said I noticed the tests fail every once in a while. I believe the edge case they're catching is the following:

- Blockchain will first write the new (reorged) blocks to db and *then* send removed/new logs. This can cause the history processor to go accidentally on the new chain before we realize there's a reorg happening.
- So we return `oldBlock1, oldBlock2, newBlock3, removeOldBlock3, newBlock3, newBlock4`. I.e. the removed logs don't match the ones we emitted for that height.

Reply:
I've pushed a fix for this. The subscription now tracks a number of block hashes that it has delivered logs for. This map is used to avoid sending removed logs for a block we haven't delivered as well as sending duplicate logs.